src/wp-includes/blocks/image.php

```diff
@@ -30,14 +31,280 @@ function render_block_core_image( $attributes, $content ) {
 		$processor->set_attribute( 'data-id', $attributes['data-id'] );
 	}
 
+	$lightbox_enabled  = false;
+	$link_destination  = isset( $attributes['linkDestination'] ) ? $attributes['linkDestination'] : 'none';
+	$lightbox_settings = block_core_image_get_lightbox_settings( $block->parsed_block );
+
+	// If the lightbox is enabled and the image is not linked, flag the lightbox to be rendered.
+	if ( isset( $lightbox_settings ) && 'none' === $link_destination ) {
+
+		if ( isset( $lightbox_settings['enabled'] ) && true === $lightbox_settings['enabled'] ) {
+			$lightbox_enabled = true;
+		}
+	}
+
+	// If at least one block in the page has the lightbox, mark the block type as interactive.
+	if ( $lightbox_enabled ) {
+		$block->block_type->supports['interactivity'] = true;
+	}
+
+	// Determine whether the view script should be enqueued or not.
+	$view_js_file = 'wp-block-image-view';
+	if ( ! wp_script_is( $view_js_file ) ) {
+		$script_handles = $block->block_type->view_script_handles;
+
+		// If the script is not needed, and it is still in the `view_script_handles`, remove it.
+		if ( ! $lightbox_enabled && in_array( $view_js_file, $script_handles, true ) ) {
+			$block->block_type->view_script_handles = array_diff( $script_handles, array( $view_js_file ) );
+		}
+		// If the script is needed, but it was previously removed, add it again.
+		if ( $lightbox_enabled && ! in_array( $view_js_file, $script_handles, true ) ) {
+			$block->block_type->view_script_handles = array_merge( $script_handles, array( $view_js_file ) );
+		}
+	}
```

Comment:
This modifies the block's `view_script_handles` as needed, but where is the actual enqueuing or dequeuing of the script handled? I may not fully understand the logic order so it would be great to clarify. I just want to make sure this script is enqueued when it should be, and not enqueued when it doesn't need to be.

Can't we instead use `wp_enqueue_script()` and `wp_dequeue_script()` here?

Reply:
It's all handled correctly. Let me explain the rationale, and I will open a Trac ticket to follow up with a native handling in WordPress core. As of today, these scripts get enqueued during block rendering in `WP_Block::render`:

https://github.com/WordPress/wordpress-develop/blob/6a61084beca680a78ed581d035d76d592d4ef117/src/wp-includes/class-wp-block.php#L271-L274

The custom code used with all core blocks applies the optimization by ensuring that these view scripts are only listed for enqueuing when the block instance of a given type is going to need them. For example, in the Image block, the view script is only useful when the Lightbox feature is activated for the block. There can be multiple block instances of the same type, so the view script gets removed from the list when the functionality isn't used. However, as soon as any block instance needs it, it must be added and never removed.

The reason why we didn't go with `wp_enqueue_script()` and `wp_dequeue_script()` is that the script handle would still need to get removed from `view_script_handles` for the block type as it gets automatically registered through `block.json` and later enqueued in the code path shared in the snippet above. It would still have to be manually enqueued when it is needed.

I might be missing some more arguments discussed a few months ago when patching core blocks, but it feels more or less the same approach that I would consider a temporary solution. In the long run, I hope we can develop some sort of automatic detection based on Interactivity API directives that would inform whether a view script is necessary for the block. It could even allow us to introduce new strategies that allow loading scripts based on some triggers like: when a user hovers over a UI element or when the UI element is visible on the screen.

Reply:
@gziolo Thanks for the explanation, most of it makes sense to me. Two follow up questions:
> For example, in the Image block, the view script is only useful when the Lightbox feature is activated for the block.

I have noticed that, however there are also at least 3 blocks where it seems to be unconditional (`core/file`, `core/navigation` and `core/search` all declare `supports.interactivity: true` in `blocks-json.php`. So it seems for those there would be a more efficient way to handle. Not a blocker here, but I still think what is happening here is a workaround, also confirmed by @tellthemachines in https://github.com/WordPress/wordpress-develop/pull/5262#discussion_r1336564955.

> The reason why we didn't go with `wp_enqueue_script()` and `wp_dequeue_script()` is that the script handle would still need to get removed from `view_script_handles` for the block type as it gets automatically registered through `block.json` and later enqueued in the code path shared in the snippet above.

How does that work for classic themes? The blocks are rendered procedurally there only after `wp_head` and `wp_enqueue_scripts`, so I believe for those themes the block scripts are already enqueued before any blocks are rendered? Please correct me if I'm wrong.

Reply:
I'm not the best person to respond to the part about `supports.interactivity: true`. It's something new that I have no knowledge about. I definitely agree with you that the current approach isn't the most efficient, and the ergonomics for the block author are far from ideal. I'm about to open a ticket so we can discuss it further and work during the 6.5 cycle on a proper way to enqueue view scripts using Interactive API only when they are needed on the front end.

> How does that work for classic themes? The blocks are rendered procedurally there only after wp_head and wp_enqueue_scripts, so I believe for those themes the block scripts are already enqueued before any blocks are rendered? Please correct me if I'm wrong.

As far as I'm aware, it should work exactly the same, because we only enqueue view scripts during block rendering with the snippet presented in my previous comment. The only difference might be that scripts are printed at the end of HTML's body because it's too late for the HTML's head.

Reply:
Ticket: https://core.trac.wordpress.org/ticket/59462.